[AD] put autodiff assoc fns in witness table

marcrasi opened the pull request:
The overall effect of this PR is to put `autodiff_associated_function` entries in the witness table for protocol function requirements that are marked `@differentiable`. See `test/AutoDiff/witness_table.swift` for a detailed example.

Here's a summary of the changes involved:

## AutoDiff changes

* Added a new `AutoDiffAssociatedFunctionIdentifier` type to act as an index for these new entries.
* Fixed `AnyFunctionType::getAutoDiffAssociatedFunctionType` to work when parameters/results are generic (conforming to `Differentiable`).
* Added `selfUncurried` flag to various functions. See the comments for explanation. This is used in thunk generation.
* `SILFunctionType::getAutoDiffAssociatedFunctionType`: taught it about `parameterIndices` and added `isMethod` flag. This stuff is used in thunk generation.

## SILWitnessTable

* Added a new `AutoDiffAssociatedFunction` kind for this new kind of entry.
* Added printing and parsing for the new kind of entry.

## SILWitnessVisitor

* Made it call a new visitor method `addAutoDiffAssociatedFunction` when it sees a function requirement with `@differentiable`.

## SILGenWitnessTable and witness thunk generation

* Taught `SILGenWitnessTable` how to `addAutoDiffAssociatedFunction`.
* Added a new `AutoDiffAssociatedFunctionIdentifier *autoDiffFuncId` parameter to a bunch of thunk-generating functions. When it's `nullptr`, the function does the usual thing. When it's non-null, the function generates a thunk for the autodiff associated function instead. I think it would be cleaner to have separate functions for these things. That would involve some refactoring, and I think it would be easiest to defer this until we upstream these changes to `master` because I might have to throw away any refactoring that I do now if it conflicts with a lot of changes in `master`.

## IRGen

* Added new kind of `WitnessTableEntry` for this new kind of entry.
* Taught `WitnessTableLayout` and `WitnessTableBuilder` how to `addAutoDiffAssociatedFunction`

Commit:
put autodiff assoc fns in witness table

slavapestov commented on lib/SILGen/SILGen.cpp:
```diff
@@ -763,6 +763,31 @@ void SILGenModule::emitAbstractFuncDecl(AbstractFunctionDecl *AFD) {
+                AutoDiffAssociatedFunctionKind::JVP,
+                diffAttr->getCheckedParameterIndices(),
+                AFD->getASTContext()));
+        emittedFunctions[declRef] = getFunction(SILDeclRef(jvpFnDecl),
```
Maybe use insert() instead of assigning to the result of operator[] and assert that it was inserted?

slavapestov reviewed (commented):
I'm a bit unhappy with the cross-cutting nature of this change since it's adding a new field to SILDeclRef. Is there another approach? I'd be happy to help brainstorm.

slavapestov commented:
Also I kindly request that for changes that touch core abstractions like this one, you request a review from a member of our team. It will be easier to iterate on the design at this time in the life cycle than when tensorflow work is sent upstream.

marcrasi commented:
Thanks for the comments, and I'll be sure to include a swift team member on changes that touch core abstractions in the future.

Your comment about adding an "addAutoDiffAssociatedFunction() visitor method" gives me an idea for an alternative that should achieve what I'm trying to do without modifying SILDeclRef.

Specifically, I could:
* Add `addAutoDiffAssociatedFunction` visitor method to SILWitnessVisitor.
* Implement `addAutoDiffAssociatedFunction` in `class SILGenWitnessTable : public SILWitnessVisitor` by generating a witness thunk for the autodiff associated function. I should be able to reuse most of the existing thunk-generating code.
* Add a new kind of `WitnessTableEntry` for autodiff associated functions. It'll have a pointer to the `AutoDiffAssociatedFunctionIdentifier`. Add some SIL printing/parsing logic for this.
* Implement `addAutoDiffAssociatedFunction` in `class WitnessTableLayout : public SILWitnessVisitor` by appending the new kind of `WitnessTableEntry` to the `Entries`.

This should get the autodiff associated functions into the witness table, which is what I want.

Do you think that is a good approach?

A possible drawback is that we won't be able to write a `witness_method` inst that refers directly to one of the associated autodiff functions, because `witness_method` refers to things by `SILDeclRef`. I think this won't actually be a problem though. @rxwei's work should enable other ways of getting at these functions. (Specifically, we'll make `witness_method <original fn>` have the ability to return a bundle of the original fn plus the autodiff associated functions).

slavapestov commented:
@marcrasi That sounds reasonable. It would make sense to have a separate code path for mangling the autodiff witness thunk too. Note that witness thunk emission already handles two cases (default and conformance) so adding another way of calling that code should not be a big problem.

In general, for SIL reviews you can tag @jckarter or @gottesmm and for IRGen @rjmccall. I'm also happy to review anything.

marcrasi commented:
I implemented the new approach, and updated the PR description to describe the new approach. @slavapestov, could you take another look? And let me know if there are any pieces that I should ask someone else to look at?

rjmccall commented on include/swift/SIL/SILWitnessTable.h:
```diff
@@ -89,7 +103,9 @@ class SILWitnessTable : public llvm::ilist_node<SILWitnessTable>,
     AssociatedTypeProtocol,
-    BaseProtocol
+    BaseProtocol,
+    // SWIFT_ENABLE_TENSORFLOW
```
I can understand why you want these comments in your fork, maybe, but please strip them when upstreaming.

rjmccall replied:
Wait, no, this patch is just going into your fork.  Why do you want these comments?  It seems like nothing but ugly bookkeeping.

rxwei replied:
It's a tradition/requirement started by Chris when he made the fork. I expect to merge AD stuff upstream ASAP because it's becoming increasingly invasive, possibly starting at the EOY.

rjmccall commented on include/swift/SIL/SILWitnessTable.h:
```diff
@@ -89,7 +103,9 @@ class SILWitnessTable : public llvm::ilist_node<SILWitnessTable>,
-    BaseProtocol
+    BaseProtocol,
+    // SWIFT_ENABLE_TENSORFLOW
+    AutoDiffAssociatedFunction
```
I was about to leave a comment here that it seems better to make this an ordinary method that's keyed by a `SILDeclRef` that describes the relationship to the original method, and then I discovered that Slava specifically told you not to do it that way.  Maybe we need to talk about this.

I'm not sure we care about the *size* of `SILDeclRef`, and the *complexity* hit seems low, especially compared to adding new kinds of witness table entries and so forth.

rxwei replied:
I'll appreciate it if you and @slavapestov et al reach a consensus and let us know! I'll also try to schedule a discussion offline about this.

marcrasi replied:
While I was trying it the `SILDeclRef` way, I noticed some difficulties:
* I'm not sure if "autodiff associated function" is completely orthogonal to all the kinds and fields in `SILDeclRef`. e.g. what if we want separate "autodiff associated functions" for `Allocator` and `Initializer`? I haven't actually thought this through and found concrete problems -- just a worry that I have.
* `WitnessTableEntry`s in `ProtocolInfo.h` are keyed by the `Decl*`, so they break when there are multiple `SILDeclRef`s in the witness table with the same underlying `Decl*`. This was pretty easy to fix by adding more data to the `WitnessTableEntry` key. But maybe there are other places that I have not yet found that also assume that all their `SILDeclRef`s have distinct `Decl*`s.

I don't have a strong opinion about whether we should do it one way or the other. Both seem to work. Just letting you know about some things that I ran into while doing it with `SILDeclRef`.

marcrasi replied:
Oh also a third difficulty:
* A few of places assume that the `SILDeclRef` refers to something with the same type as the underlying `Decl*`. Adding "autodiff associated functions" breaks this assumption. It was pretty easy to adjust all the places that I found that mattered for this PR, but I'm not sure if there are other places lurking that are harder to find or use the assumption in more subtle ways.

rjmccall replied:
Do you remember an example of such a place?  That would be very surprising since e.g. foreign and non-foreign `SILDeclRef`s obviously have different types.

rjmccall replied:
> * I'm not sure if "autodiff associated function" is completely orthogonal to all the kinds and fields in `SILDeclRef`. e.g. what if we want separate "autodiff associated functions" for `Allocator` and `Initializer`? I haven't actually thought this through and found concrete problems -- just a worry that I have.

I don't think it should be orthogonal.  I wouldn't add a new kind at all: you should add the ability to add a derivation path to an arbitrary other `SILDeclRef`.

> * `WitnessTableEntry`s in `ProtocolInfo.h` are keyed by the `Decl*`, so they break when there are multiple `SILDeclRef`s in the witness table with the same underlying `Decl*`. This was pretty easy to fix by adding more data to the `WitnessTableEntry` key. But maybe there are other places that I have not yet found that also assume that all their `SILDeclRef`s have distinct `Decl*`s.

Hmm, that's unfortunate.  I don't know why they would need to break down the `SILDeclRef` like that.

Anyway, Slava and I will talk about it.

jckarter replied:
The `WitnessTableEntry` thing is probably just divergent evolution we never patched up when SIL was added in. It should almost definitely just use `SILDeclRef` now.

marcrasi replied:
> Do you remember an example of such a place? That would be very surprising since e.g. foreign and non-foreign SILDeclRefs obviously have different types.

Some places are:
* Search `getDecl()->getInterfaceType()` in [SILPrinter.cpp](https://github.com/apple/swift/blob/01bfdc162e2bfdaa609e25a30c3b6b965fccd272/lib/SIL/SILPrinter.cpp).
* [This lookup logic in ParseSIL.cpp](https://github.com/apple/swift/blob/master/lib/ParseSIL/ParseSIL.cpp#L1928).

Actually, I think that is all that I found. Maybe it does make sense to do it that way in SILPrinter/ParseSIL if you think of the type in the printed SIL as always being the type of the underlying decl.

rjmccall replied:
No, those are both correct.  The type there is not the type of the constant identified by the `SILDeclRef` — it's the type of the declaration that serves as the root of the `SILDeclRef`, and it's only there to potentially resolve between overloads.

marcrasi replied:
That makes sense. So the third difficulty I listed is not actually a problem.

rjmccall replied:
Slava and I talked about this after the meeting today and decided that the `SILDeclRef` approach makes sense.  The two kinds of derivative should each be `SILDeclRef::Kind`s, and you should be able to re-use one of the other fields (like `defaultArgIndex`) for the order of the derivative.

...it's quite unfortunate that "higher-order" is used for both repeated derivation and functions of functions.

rxwei replied:
Awesome. Thanks John!

marcrasi commented on lib/SILGen/SILGenThunk.cpp:
```diff
@@ -268,6 +268,35 @@ SILValue SILGenFunction::emitGlobalFunctionRef(SILLocation loc,
 }
 
+// SWIFT_ENABLE_TENSORFLOW
+SILValue SILGenFunction::emitGlobalAutoDiffAssociatedFunctionRef(
```
oh cool, that is very clean.

it seems that `autodiff_function_extract` is not calculating the right type right now, which is blocking me from adding this to this PR, so I'll keep the lookup logic for now but add a todo to replace it with `autodiff_function_extract`.

marcrasi commented on lib/SILGen/SILGenType.cpp:
```diff
@@ -327,7 +327,8 @@ template<typename T> class SILGenWitnessTable : public SILWitnessVisitor<T> {
                                                   requirementRef.kind),
-                                       witness);
+                                       // SWIFT_ENABLE_TENSORFLOW
+                                       witness, nullptr);
```
done

rxwei reviewed (approved):
LGTM.

Re @slavapestov @rjmccall @jckarter : We've decided to land this patch in the tensorflow branch to unblock things. Will discuss with you all before upstreaming.

bgogul commented:
> Re @slavapestov @rjmccall @jckarter : We've decided to land this patch in the tensorflow branch to unblock things. Will discuss with you all before upstreaming.

@slavapestov @rjmccall @jckarter: Just to clarify a bit more. We would like to continue the discussion and converge on the right design **now**. As you guys suggested, it is not a good idea to wait until we upstream to discuss such core changes. What Richard meant was that, if we do not converge in the next two days, we would like to submit the patch on the tensorflow branch to unblock the rest of the team. @marcrasi  and @rxwei  will continue working with you guys to converge on the design.

marcrasi commented:
@swift-ci please test tensorflow

marcrasi commented:
@swift-ci please test tensorflow
